File: packages/core-abstractions/src/base/BaseEntity.ts
 
/**
 * Base Entity class following DDD principles
 * All domain entities should inherit from this class
 */
export abstract class BaseEntity<TId = string> {
  public readonly id: TId;
  public readonly createdAt: Date;
  public updatedAt: Date;
  private _domainEvents: DomainEvent[] = [];

  constructor(id: TId) {
    this.id = id;
    this.createdAt = new Date();
    this.updatedAt = new Date();
  }

  /**
   * Compares entities by their IDs
   */
  public equals(other: BaseEntity<TId>): boolean {
    if (other == null || other.constructor !== this.constructor) {
      return false;
    }
    return this.id === other.id;
  }

  /**
   * Adds a domain event to be dispatched
   */
  protected addDomainEvent(event: DomainEvent): void {
    this._domainEvents.push(event);
  }

  /**
   * Gets and clears all domain events
   */
  public getDomainEvents(): DomainEvent[] {
    const events = [...this._domainEvents];
    this.clearDomainEvents();
    return events;
  }

  /**
   * Clears all domain events
   */
  public clearDomainEvents(): void {
    this._domainEvents = [];
  }
}

/**
 * Base Domain Event interface
 */
export interface DomainEvent {
  readonly eventId: string;
  readonly occurredOn: Date;
  readonly eventType: string;
  readonly aggregateId: string;
}

File: backend/Modules/01-Core.Module/Domain/Common/BaseEntity.cs
using System.ComponentModel.DataAnnotations.Schema;
using Core.Module.Domain.Common;

namespace Core.Module.Domain.Common
{
    public abstract class BaseEntity
    {
        public int Id { get; protected set; }
        
        private readonly List<DomainEvent> _domainEvents = new();
        
        [NotMapped]
        public IReadOnlyCollection<DomainEvent> DomainEvents => _domainEvents.AsReadOnly();

        public void AddDomainEvent(DomainEvent domainEvent)
        {
            _domainEvents.Add(domainEvent);
        }

        public void RemoveDomainEvent(DomainEvent domainEvent)
        {
            _domainEvents.Remove(domainEvent);
        }

        public void ClearDomainEvents()
        {
            _domainEvents.Clear();
        }
    }
}